File: app/game/Controller.java
package game;

import java.util.HashMap;
import java.util.Map;

/**
 * Created by krotkov on 7/30/2016.
 */
public class Controller {

    private Map<Long, SingleGame> games;

    Controller() {
        games = new HashMap<>();
    }

    public void processMessage(long chatId, User user, String text) {
        if (text.equals("start") && !games.containsKey(chatId)) {
            startGame(chatId);
            return;
        }

        if (!games.containsKey(chatId)) {
            return;
        }

        games.get(chatId).processMessage(user, text);
    }

    public void startGame(long chatId) {
        games.put(chatId, new SingleGame(new BotImpl(chatId, this)));
    }

    public void finishGame(long chatId) {
        games.remove(chatId);
    }
}

File: app/game/Controller.scala
package game

import java.util.concurrent.ConcurrentHashMap

import bot.ChatImpl
import telegram.TelegramApi

import scala.concurrent.Future

class Controller(api: TelegramApi) {

  private val games = new ConcurrentHashMap[Long, SingleGame]()

  def processMessage(chatId: Long, user: User, text: String): Future[Unit] = {
    Option(games.get(chatId)) match {
      case Some(game) =>
        System.err.println("Game already exists")
        game.processMessage(user, text)
      case None =>
        System.err.println("Game doesn't exist")
        if (text.startsWith("start")) {
          startGame(chatId, text.substring(6))
        }
    }
    Future.successful(())
  }

  def startGame(chatId: Long, cats: String) {
    System.err.println("game started");
    games.put(chatId, new SingleGame(new ChatImpl(chatId, this, api), cats))
  }

  def finishGame(chatId: Long) {
    System.err.println("finishing game in Controller");
    games.remove(chatId)
  }

}
